ĐÂY LÀ NHÁP THÔI (XÓA ĐI NHÉ)
+ ============ Method 1 ===================
1. Khám phá dữ liệu: có bao nhiêu dòng/cột, ý nghĩa dòng/cột, trùng dòng/cột, kiểu DL có phù hợp

2. Tiền xử lý dữ liệu: chuyển dữ liệu về kiểu dữ liệu phù hợp 

3. Khám phá dữ liệu (phân bổ): phân bố dữ liệu của từng cột như thế nào (chia theo kiểu DL), min/max/median/missing-ratio

+ ============ Method 2 ===================
1. Loại trùng các dòng 

2. Khám phá dữ liệu: dữ liệu có bao nhiêu dòng/cột, bỏ cột/dòng trùng, ý nghĩa dòng/cột, kiểm tra kiểu dữ liệu các cột

3. Tiền xử lý dữ liệu: chia thành kiểu DL category và numerical, tính min/max/median/missing-ratio, 

4. Chuẩn hóa cột: chuyển đổi kiểu dữ liệu sao cho phù hợp 

test_df = pd.read_csv('../data/test.csv', converters={'Genres' : pd.eval, 'Themes' : pd.eval, 'Type Review' : pd.eval})
test_df.head(2)

# Data Pre-Processing and Exploration

## Import Libraries

These are the main libraries used for data exploration:
+ `numpy`: 
+ `pandas`:
+ `matplotlib`

In [1]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
import ast
import re
import datetime as dt
pd.options.mode.chained_assignment = None

## Exploring Data (rows, cols, dtype)

### Read data from file `.csv`

In [2]:
manga_df = pd.read_csv('../data/comic.csv')

In [3]:
manga_df.head()

Title  Score    Vote  Ranked  \
0                                          Berserk   9.47  330372       1   
1  JoJo no Kimyou na Bouken Part 7: Steel Ball Run   9.30  156011       2   
2                                         Vagabond   9.24  135878       3   
3                                        One Piece   9.22  365947       4   
4                                          Monster   9.15   93668       5   

   Popularity  Members Favorite  Volumes Chapters      Status  \
0           1  663,681  122,589  Unknown  Unknown  Publishing   
1          26  255,574   42,791       24       96    Finished   
2          15  363,767   40,004       37      327   On Hiatus   
3           3  598,004  114,319  Unknown  Unknown  Publishing   
4          29  235,740   20,468       18      162    Finished   

                        Published  \
0              Aug  25, 1989 to ?   
1  Jan  19, 2004 to Apr  19, 2011   
2   Sep  3, 1998 to May  21, 2015   
3              Jul  22, 1997 to ?   
4   Dec  5, 1994 to Dec  20, 2001   

                                              Genres  \
0  ['Action', 'Adventure', 'Award Winning', 'Dram...   
1  ['Action', 'Adventure', 'Mystery', 'Supernatur...   
2           ['Action', 'Adventure', 'Award Winning']   
3                 ['Action', 'Adventure', 'Fantasy']   
4              ['Award Winning', 'Drama', 'Mystery']   

                                              Themes  \
0  ['Gore', 'Military', 'Mythology', 'Psychologic...   
1                                                 []   
2                          ['Historical', 'Samurai']   
3                                                 []   
4                    ['Adult Cast', 'Psychological']   

                                              Author  Total Review  \
0   Miura, Kentarou (Story & Art), Studio Gaga (Art)           259   
1                      Araki, Hirohiko (Story & Art)           128   
2  Inoue, Takehiko (Story & Art), Yoshikawa, Eiji...            95   
3                        Oda, Eiichiro (Story & Art)           206   
4                       Urasawa, Naoki (Story & Art)            76   

     Type Review  
0  [234, 15, 10]  
1    [120, 7, 1]  
2     [87, 7, 1]  
3  [173, 17, 16]  
4     [64, 7, 5]

### Total number of rows and columns of data

In [4]:
n_rows , n_cols = manga_df.shape
print(f'Data has {n_rows} rows and {n_cols} columns')

Data has 5000 rows and 16 columns


### Meaning of each row and each column

+ Each rows represents information about a manga series

+ The following is the meaning of each column in the data

  | **COLUMN**            | **MEANING**                                                                                     |
  |:----------------------|:------------------------------------------------------------------------------------------------|
  |**`Title`**            | Title of the manga (written in English phonetic)                                                |
  |**`Score`**            | Score on the MyAnimeList site (MAL)                                                             |
  |**`Vote`**             | Number of readers voting for the manga                                                          |
  |**`Ranked`**           | Ranking of manga on the web MyAnimeList (MAL)                                                   |
  |**`Popularity`**       | The popularity of the manga                                                                     |
  |**`Members`**          | Number of readers who have this manga in their list                                             |
  |**`Favorite`**         | Number of readers who love this manga                                                           |
  |**`Volumes`**          | Volume number of the manga                                                                      |
  |**`Chapters`**         | Chapter number of the manga                                                                     |
  |**`Status`**           | Status of the manga (ongoing, completed, on hiatus,...)                                         |
  |**`Published`**        | Release time to the end time of the manga                                                       |
  |**`Genres`**           | Genres of manga                                                                                 |
  |**`Themes`**           | The themes of the manga                                                                         |
  |**`Author`**           | Author information of the manga                                                                 |
  |**`Total Review`**     | Number of readers leaving comments on the manga                                                 |
  |**`Type Review`**      | Number of readers for each comment category (Recommended / Mixed feeling / Not recommended)     |
  

### Check and handle duplicates

In [5]:
duplicated_rows = manga_df.duplicated().sum()

if duplicated_rows == 0:
    print('There are 0 rows that are duplicated, which means each row in the DataFrame is unique.')
    print('So that we do not need to continue processing duplicate lines')
else:
    print(f'There are {duplicated_rows} rows that are duplicated so we need to drop those {duplicated_rows} rows')
    manga_df = manga_df.drop_duplicates()

There are 1 rows that are duplicated so we need to drop those 1 rows


### Check and handle data type of columns

Next, we calculate the data type (dtype) of each column in DataFrame `manga_df` and save the result into Series `dtypes` (This Series has the index as the column name).

In [6]:
dtypes = manga_df.dtypes
dtypes

Title            object
Score           float64
Vote              int64
Ranked            int64
Popularity        int64
Members          object
Favorite         object
Volumes          object
Chapters         object
Status           object
Published        object
Genres           object
Themes           object
Author           object
Total Review      int64
Type Review      object
dtype: object

To better understand columns of type `object`, we will analyze them into each subtype using function `open_object_dtype()`

In [7]:
def open_object_dtype(s):
    dtypes = set()
    dtypes.update(s.apply(type))
    return dtypes

Now, considering columns whose data type is object more clearly

In [8]:
manga_df[manga_df.dtypes.index[manga_df.dtypes == object]].apply(open_object_dtype)

Title                           {<class 'str'>}
Members                         {<class 'str'>}
Favorite                        {<class 'str'>}
Volumes                         {<class 'str'>}
Chapters                        {<class 'str'>}
Status                          {<class 'str'>}
Published                       {<class 'str'>}
Genres                          {<class 'str'>}
Themes                          {<class 'str'>}
Author         {<class 'str'>, <class 'float'>}
Type Review                     {<class 'str'>}
dtype: object

Let's see which columns have appropriate data types and which columns do not

+ Suitability: `Title`, `Score`, `Vote`, `Ranked`, `Popularity`, `Status` and `Total Review`

+ Need to convert to **int**: `Members`, `Favorite`, `Volumes` and `Chapters`

+ Need to convert to **list**: `Type Review`, `Genres` and `Themes`

+ Need to convert to **str**: `Author` 

+ Need to convert to **datetime**: `Published`

In [9]:
# Ko phù hợp: 
#     + Type Review (ob -> list)
#     + Member, favorite, volume, chapter,   (ob -> int)
#     + Status, Genres, Themes, Author (ob -> str)
#     + Published: chia thanh 2 cot ngay phat hanh va ngay ket thuc (ob -> datetime)

## Pre-processing data

### Normalizing columns (convert the data types of columns accordingly)

Firstly, converting the data type of the columns to **int / float**

In [10]:
# Convert to int
manga_df['Members'].replace(',','', regex=True, inplace=True)
manga_df['Favorite'].replace(',','', regex=True, inplace=True)
manga_df['Members'] = pd.to_numeric(manga_df['Members'], errors='coerce')
manga_df['Favorite'] = pd.to_numeric(manga_df['Favorite'], errors='coerce')

# Convert to float because type int does not contain NaN values
manga_df['Volumes'] = pd.to_numeric(manga_df['Volumes'], errors='coerce')
manga_df['Chapters'] = pd.to_numeric(manga_df['Chapters'], errors='coerce')


Let's check again data type of `Members`, `Favorite`, `Volumes` and `Chapters`

In [11]:
check = ['Members', 'Favorite', 'Volumes', 'Chapters']
manga_df[check].dtypes

Members       int64
Favorite      int64
Volumes     float64
Chapters    float64
dtype: object

Secondly, converting the data type of the columns to **list**

In [12]:
manga_df[['Genres', 'Themes', 'Type Review']] = \
    manga_df[['Genres', 'Themes', 'Type Review']].apply(lambda col : col.apply(ast.literal_eval), axis=1)

+ Looking at the `Type Review` column, we see that the data type of this column is quite difficult to access (the column's data type is a list of 3 numbers representing the number of comments of three comment types: `Recommended`, `Mixed feelings` and `Not recommended`) so we will divide `Type Review` into 3 new columns: `Recommended`, `Mixed Feelings` and `Not Recommended`. 
+ After dividing into 3 new columns, the `Type Review` column will be removed from the data. 
+ Additionally, the `Total Review` column which is the sum of those 3 new columns will also be removed

In [13]:
manga_df['Recommended'] = 0
manga_df['Mixed Feelings'] = 0
manga_df['Not Recommended'] = 0

for row, i in zip(manga_df['Type Review'], manga_df.index):
    manga_df['Recommended'][i] = row[0]
    manga_df['Mixed Feelings'][i] = row[1]
    manga_df['Not Recommended'][i] = row[2]

Before deleting column `Total Review`, we check whether the data collected is reasonable

In [14]:
def checking_reviews(df: pd.DataFrame) -> bool:
    return df["Total Review"].equals((df["Recommended"] + df["Mixed Feelings"] + df["Not Recommended"]))

In [15]:
checking_reviews(manga_df)==True

True

Because the `Total Review` column is the sum of the 3 columns we just created (checked to be True above), we will delete the `Total Review` column from the data.

In [16]:
manga_df = manga_df.drop(columns=['Type Review', 'Total Review'])

Let's check the data type of the 3 newly created columns

In [17]:
check = ['Recommended', 'Mixed Feelings', 'Not Recommended']
manga_df[check].dtypes

Recommended        int64
Mixed Feelings     int64
Not Recommended    int64
dtype: object

Thirdly, converting the data type of the columns to **datetime**

Because the `Published` column is quite difficult to store during discovery and continuing with the following steps (the data in the column is `release date to end date`), we will split the `Published` column into 2 new columns. is column `Release date` and column `Completed date` and remove column `Published`

In [18]:
start = []
end = []
for row in manga_df['Published']:
    try:
        s, e = row.split(' to ')
        start.append(s)
        end.append(e)
    except:
        s = np.nan
        e = np.nan
        start.append(s)
        end.append(e)

manga_df['Release date'] = pd.DataFrame(start)
manga_df['Completed date'] = pd.DataFrame(end)

manga_df['Release date'] = pd.to_datetime(manga_df['Release date'], format='%b %d, %Y', errors='coerce')
manga_df['Completed date'] = pd.to_datetime(manga_df['Completed date'], format='%b %d, %Y', errors='coerce')

manga_df = manga_df.drop(columns='Published')

Let's data type of `Release date` and `Completed date`

In [19]:
check = ['Release date', 'Completed date']
manga_df[check].dtypes

Release date      datetime64[ns]
Completed date    datetime64[ns]
dtype: object

Finally, converting the data type of the columns to **string**

In [20]:
# Converted to str, any value that is str remains the same and any value that is NaN is converted to an empty string
manga_df['Author'] = manga_df['Author'].replace(to_replace=np.nan, value='')

Now, checking all data type columns after conversion

In [21]:
manga_df.apply(open_object_dtype)

Title                                                {<class 'str'>}
Score                                              {<class 'float'>}
Vote                                                 {<class 'int'>}
Ranked                                               {<class 'int'>}
Popularity                                           {<class 'int'>}
Members                                              {<class 'int'>}
Favorite                                             {<class 'int'>}
Volumes                                            {<class 'float'>}
Chapters                                           {<class 'float'>}
Status                                               {<class 'str'>}
Genres                                              {<class 'list'>}
Themes                                              {<class 'list'>}
Author                                               {<class 'str'>}
Recommended                                          {<class 'int'>}
Mixed Feelings                    

## Exploring data (distribution of values)

### Distribution of numeric columns

Columns have numeric data types: 
+ `Score`
+ `Vote`
+ `Ranked`
+ `Popularity`
+ `Members`
+ `Favorite`
+ `Volumes`
+ `Chapters`
+ `Recommended`
+ `Mixed Feelings`
+ `Not Recommended`

For columns with numeric data types, we will calculate:
- Percentage (from 0 to 100) of missing values
- The min
- The lower quartile 
- The median 
- The upper quartile 
- The max

We will save the results to a DataFrame `num_col_dist_df`, where:
- The names of the columns are the names of the numeric columns in `manga_df`
- Names of rows: "missing_ratio", "min", "lower_quartile", "median", "upper_quartile", "max"  

For ease of viewing, we round all values to 1 decimal place using the `.round(1)` method.

In [22]:
num_col_dist_df = manga_df[['Score','Vote','Ranked','Popularity','Members','Favorite','Volumes','Chapters','Recommended','Mixed Feelings','Not Recommended']]

def missing_ratio(s):
    return (s.isna().mean() * 100).round(1)

def median(df):
    return (df.quantile(0.5)).round(1)

def lower_quartile(df):
    return (df.quantile(0.25)).round(1)

def upper_quartile(df):
    return (df.quantile(0.75)).round(1)

num_col_dist_df = num_col_dist_df.agg([missing_ratio, "min", lower_quartile, median, upper_quartile, "max"])
num_col_dist_df

Score      Vote  Ranked  Popularity   Members  Favorite  \
missing_ratio    0.00       0.0     0.0         0.0       0.0       0.0   
min              7.27     102.0     1.0         1.0     219.0       0.0   
lower_quartile   7.40     540.0  1250.5      1643.0    2042.0      19.0   
median           7.60    1458.0  2501.0      4201.0    4615.0      55.0   
upper_quartile   7.80    3999.5  3750.5      8503.5   11616.5     211.0   
max              9.47  399972.0  5052.0     31592.0  663681.0  122589.0   

                Volumes  Chapters  Recommended  Mixed Feelings  \
missing_ratio      27.3      27.5          0.0             0.0   
min                 1.0       1.0          0.0             0.0   
lower_quartile      2.0      13.0          1.0             0.0   
median              6.0      34.0          2.0             0.0   
upper_quartile     12.0      80.0          5.0             1.0   
max               200.0    1957.0        280.0            56.0   

                Not Recommended  
missing_ratio               0.0  
min                         0.0  
lower_quartile              0.0  
median                      0.0  
upper_quartile              0.0  
max                       112.0

As we can see:
+ About missing percentage:
      + All columns except `Volumes` and `Chapters` have full data
      + The column `Volumes` has **27.3%** of the total values missing. While the `Chapters` column is **27.5%**. It can be seen that the missing rate of these two columns is not too high and is acceptable. Columns `Volumes` and `Chapters` are missing mainly due to their `Status` (incomplete, on hold, on hiatus,...), because these two columns only display volume and chapter numbers when that set has ended (completed or stopped indefinitely).
     
+ About min, median, lower_quartile, upper_quartile and max:
     + Except for the `Score` column, numeric columns all have very wide ranges

In [23]:
# Missing manga volume number column grouped by status
np.unique(manga_df['Status'][manga_df['Volumes'].isna()], return_counts=True)

(array(['Discontinued', 'Finished', 'On Hiatus', 'Publishing'],
       dtype=object),
 array([   1,  210,    4, 1151], dtype=int64))

In [24]:
# Missing manga chapter number column grouped by status
np.unique(manga_df['Status'][manga_df['Chapters'].isna()], return_counts=True)

(array(['Discontinued', 'Finished', 'On Hiatus', 'Publishing'],
       dtype=object),
 array([   1,  209,   14, 1150], dtype=int64))

+ As mentioned above, the manga is missing chapters or volumes because it is not yet completed and are in progress (Publishing). `Chapters` and `Volumes` will be updated when the manga is completed or discontinued.

+ For manga that only have 'Chapters' but no 'Volumes', the biggest reason is one-shot manga series (series with only 1 chapter) so there will be no volumes. Such one-shot manga make up a large proportion on MAL.

+ For manga that only have 'Volumes' but no 'Chapters', the biggest reason is that light novels (light_novels or novels) are often published in volumes and the author does not divide chapters when writing. There are also many novel series on MAL.

Next, after identifying the basic statistical numbers that describe our data, we further need to determine the features that have a large number of missing values. Such features are not useful for the analysis stage and must be removed from the dataset.

We decided that if the percentage of missing values is greater than 75%, the column is dropped from the dataframe and an updated dataframe is returned.

In [25]:
def drop_missing_features(df: pd.DataFrame, missing_lst: dict = dict(num_col_dist_df.iloc[0]), threshold: float = 75.0) -> pd.DataFrame:
    if (df is None) or (missing_lst is None) or (threshold is None):
        print(f"[LOG] Invalid.")
        raise ValueError
    
    # Create a copy of the dataframe to avoid changing the original
    df_cp = df.copy()
    
    # List to maintain the columns to drop
    cols_to_trim = []
    
    for key, value in missing_lst.items():
        if float(value) > threshold:
            cols_to_trim.append(key)
            
    if len(cols_to_trim) > 0:
        # drop the columns identified using the dataframe drop() method
        df_cp = df_cp.drop(columns=cols_to_trim)
        print("[LOG] Dropped columns: " + " ".join(cols_to_trim))
    else:
        print("[LOG] Have no columns to trim.")
        
    return df_cp

In [26]:
manga_df = drop_missing_features(manga_df, dict(num_col_dist_df.iloc[0]))

[LOG] Have no columns to trim.


After remove features which have large missing values, our dataframe still have missing values. So that, we need to fill these missing values so that they can be used in analysis.

We fill these missing values with the **median** of the column values.

In [27]:
def filling_missing_value(df: pd.DataFrame) -> pd.DataFrame:
    if (df is None):
        print(f"[LOG] Invalid.")
        raise ValueError
    
    # Create a copy of the dataframe to avoid changing the original
    df_cp = df.copy()
    
    # Get the list of columns in the dataframe
    cols_list = list(df_cp.columns)
    
    # Exclude the last column - Country, Year
    cols_list.remove("Title")
    cols_list.remove("Status")
    cols_list.remove("Genres")
    cols_list.remove("Themes")
    cols_list.remove("Author")
    cols_list.remove("Release date")
    cols_list.remove("Completed date")
    
    # Replace all None values with NaN, fillna only works on nans
    df_cp.fillna(value=np.nan, inplace=True)
    
    # Replace all NaN values with the median of the column values
    for col in cols_list:
        df_cp[col].fillna((df_cp[col].median()), inplace=True)
        
    return df_cp

In [28]:
manga_df = filling_missing_value(df=manga_df)

## Distribution of non-numeric columns